### ISSUE TITLE
Text sent by telemetry

### ISSUE THREAD

**Issue by mrosenfield2** [FrSkyRC/ETHOS-Feedback-Community#2680]
I want to switch from Spektrum RC radios to FRSky radios, but my system requires the ability to send text messages from the remote system to the transmitter.
Spektrum has this capability - 9 lines of 18 characters can be displayed on one of the telemetry screens.
Can this be added to FRSky radios?

**Comment by bsongis-frsky** [comment]
9 line of 18 chars. How many frames are needed on Spektrum system for that?

**Comment by bsongis-frsky** [comment]
With a Lua script it will be possible already. I am thinking of the possibility to define such a sensor in the spec. without any need to implement a Lua script.

**Comment by mrosenfield2** [comment]
Each line is output as one frame. Sorry, it’s up to 13 characters per line.
I’ve attached their telemetry spec; see bottom of page 13 and top of page 14.

Michael Rosenfield
RSR Engineering, LLC
2420 E Hermosa Vista Dr.
Mesa, AZ 85213
480-861-6456
www.rsr-engineering.com

**Comment by mrosenfield2** [comment]
I am pretty new working with FrSky. How would I define a sensor, or use LUA script?

**Comment by frankiearzu** [comment]
This looks like is using the  Spektrum TextGen Frames..   Each telemetry message has a Line# + Text (13 chars).
The RX keeps broadcasting the TextGen screen repetitively line by line, so you can start collecting the lines in the middle (like line 4) , and just populate the screen incrementally.   I think with the recent changes for DSM telemetry push/pop (Oct 23) intended for Forward Programming, a lua script can do it.  Im working on a prototype for TextGen lua Tool.

This is Spektrum specific, should we add Spektrum to the title?

**Comment by frankiearzu** [comment]
Got some good coding session last night, and was able to port TextGen to work with 1.6.0 nightly builds.
Was able to do Avian Prog as well as the new LemonRX Gen2 RX programming,

Some preview:
![SmartTel-TextGen-Avian](https://github.com/user-attachments/assets/28ca767a-d62d-4470-9a08-a473b7bb5de3)
![SmartTel-TextGen-Lemon](https://github.com/user-attachments/assets/cf401dc2-9cbc-4984-9deb-d195fe57413e)

Very Early version can be downloaded from here
https://github.com/frankiearzu/DSMTools/tree/main/DSM_Telemetry/In%20Development/Ethos

**Comment by robthomson** [comment]
Nice to see this working

Would it not make sense to look at ethos forms?

You could really make this look native.

Happy to assist if you want.

**Comment by frankiearzu** [comment]
Right now im just doing a quick port from my code on EdgeTX that works on both color and B&W small screens.
My plan is to make them work first, and then we can collectively work on enhancing the UI. 

@robthomson  Definitly apreciate your help.. UI is not my forte :-)   Give me some time to stabilize the basics, and we can enhance the UI afterwards. I usually add data simulators to my code, so the UI can be developed in the Simulator quicker.

**Comment by robthomson** [comment]
For sure let me know when ready.

I have been over the learning curve with porting the rf stuff.

What will make it 100% easier is if we can emulate the api calls to the rx.  That way we can simulate the whole thing.  Not sure if your code does that?

**Comment by frankiearzu** [comment]
Yes, i usually add some kind of simulation code that can be turned on/off to be able to develop GUI on the PC-Sim. Will make sure the TextGen portion is simulated.

**Comment by frankiearzu** [comment]
@robthomson checked in a version with a simulation for the first 3-4 pages.
If you want, you can clone my github and modify the script to make it look better.
Ethos code in there, probably will be goog to start with the dsm-smart-tel script, Forward prog is still in progress.
On the top of the script, there is a "SIMULATION=true/false" and runs on 1.6.0 nightly. So far is "key" based and not touch.  
https://github.com/frankiearzu/DSMTools/tree/main/DSMTools-src/ethos/src/scripts

**Comment by robthomson** [comment]
OK.. Will have a look at it ans see what needs doing.

**Comment by robthomson** [comment]
@frankiearzu ... would love to have a quick chat if you are around?

Had a look at code and I can see what we prob need to do - but kind of think we need to do in a 100% separate branch. 

Use of forms is VERY different in approach.

I am thinking atm that what we need to do is tackle this as follows:

I provide you a 'simple' form/menu system that presents you with:

This will essetially parse a copy of tables.

- table 1.  
   This will be the menu that is displayed.  
- table 2.
   this is called when we click on the menu presented from table 1.  The content of the table will then determine how/what the sub page looks like.
   
Essentially what I would like to do is allow us to just abstract the queriying of the data into a formated table that is returned.  I then parse the table and display a menu.   Quite simple all up.

Do you use discord?  Much easier to chat on this in person and work best way forward.
